felt/src/lib.rs

```diff
@@ -108,162 +58,274 @@ macro_rules! felt_str {
+            FieldElement::zero() - FieldElement::one() - FieldElement::new(abs_minus_one)
+        };
+        Self { value }
+    }
+}
+
+from_num!(usize, u64);
+from_num!(u8, u64);
+from_num!(u16, u64);
+from_num!(u32, u64);
+
+// TODO: move to upstream?
+impl From<u64> for Felt252 {
+    fn from(value: u64) -> Self {
+        let value = FieldElement::new(UnsignedInteger::from_u64(value));
+        Self { value }
+    }
+}
+
+// TODO: move to upstream?
+impl From<u128> for Felt252 {
+    fn from(value: u128) -> Self {
+        let value = FieldElement::new(UnsignedInteger::from_u128(value));
+        Self { value }
+    }
+}
+
+// TODO: bury BigUint?
+impl From<BigUint> for Felt252 {
+    fn from(mut value: BigUint) -> Self {
+        if value >= *CAIRO_PRIME_BIGUINT {
+            value = value.mod_floor(&CAIRO_PRIME_BIGUINT);
+        }
+        let mut limbs = [0; 4];
+        for (i, l) in (0..4).rev().zip(value.iter_u64_digits()) {
+            limbs[i] = l;
+        }
+        let value = FieldElement::new(UnsignedInteger::from_limbs(limbs));
+        Self { value }
+    }
+}
+
+// TODO: bury BigInt?
+// NOTE: used for deserialization
+impl From<BigInt> for Felt252 {
+    fn from(value: BigInt) -> Self {
+        let val = value.mod_floor(&CAIRO_PRIME_BIGUINT.to_bigint().expect("cannot fail"));
+        let mut limbs = [0; 4];
+        for (i, l) in (0..4).rev().zip(val.iter_u64_digits()) {
+            limbs[i] = l;
+        }
+        let value = FieldElement::new(UnsignedInteger::from_limbs(limbs));
+        Self { value }
+    }
+}
+
+// NOTE: used for serialization
+impl From<Felt252> for BigInt {
+    fn from(value: Felt252) -> Self {
+        value.to_bigint()
+    }
+}
 
 impl Felt252 {
     pub fn new<T: Into<Felt252>>(value: T) -> Self {
         value.into()
     }
-    pub fn modpow(&self, exponent: &Felt252, modulus: &Felt252) -> Self {
-        Self {
-            value: self.value.modpow(&exponent.value, &modulus.value),
-        }
-    }
-    pub fn iter_u64_digits(&self) -> U64Digits {
-        self.value.iter_u64_digits()
+    pub fn iter_u64_digits(&self) -> impl Iterator<Item = u64> {
+        self.value.representative().limbs.into_iter().rev()
     }
 
     pub fn to_le_bytes(&self) -> [u8; 32] {
-        let mut res = [0u8; 32];
-        let mut iter = self.iter_u64_digits();
-        let (d0, d1, d2, d3) = (
-            iter.next().unwrap_or_default().to_le_bytes(),
-            iter.next().unwrap_or_default().to_le_bytes(),
-            iter.next().unwrap_or_default().to_le_bytes(),
-            iter.next().unwrap_or_default().to_le_bytes(),
-        );
-        res[..8].copy_from_slice(&d0);
-        res[8..16].copy_from_slice(&d1);
-        res[16..24].copy_from_slice(&d2);
-        res[24..].copy_from_slice(&d3);
-        res
+        // TODO: upstream should return array
+        self.value
+            .representative()
+            .to_bytes_le()
```

Comment:
This makes me worry a bit in terms of performance due to the allocation. Can't we iter the digits of the representative as it was done originally?

Reply:
Good idea!